Dear Editor,

We submit our manuscript titled,
_"jax-unirep: jax-unirep: An accelerated, performant, and user-friendly implementation of UniRep using JAX"._

In this manuscript, we detail significant usability and speed enhancements
to UniRep, a previously-published deep learning model
used for protein engineering purposes.
UniRep is capable of generating
fixed-length vector representations of any protein sequence,
and its weights can be "evolutionarily tuned" to homologous protein subsets.
Though the original UniRep implementation was a significant scientific advance,
its speed, documentation, and usability left much to be desired.
In our work, we

1. speed up the model by up to **100X over the original**,
2. implemented extensive **software testing and in-line documentation**, and
3. provided **user-friendly application programming interfaces**
that suit a computational protein engineer's workflow,
which we use in our own projects.

In its short term of existence since first release on GitHub,
jax-unirep has already acquired users at academic
and industrial research institutes around the globe,
including the Novartis Institutes for Biomedical Research and UC Berkeley
(the home institutions of the authors), Novo Nordisk, U Chicago, and more.
We believe this submission is suitable for Bioinformatics
due to its relevance to machine learning on protein sequences,
and as an Application Note because it is a significantly enhanced version
of a previously published machine learning model.
We also hope our submission to Bioinformatics as an Application Note
will help increase the impact of the model in protein engineering applications.

Sincerely,

Eric J. Ma
